hazelcast/src/main/java/com/hazelcast/internal/nearcache/NearCacheRecord.java

```diff
@@ -18,25 +18,47 @@
 
 import com.hazelcast.internal.eviction.Evictable;
 import com.hazelcast.internal.eviction.Expirable;
+import com.hazelcast.map.impl.record.Record;
 
 import java.util.UUID;
 
+import static com.hazelcast.internal.util.TimeUtil.zeroOutMs;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
 /**
- * An expirable and evictable data object which represents a Near Cache entry.
- * <p>
+ * An expirable and evictable data object
+ * which represents a Near Cache entry.
+ *
  * Record of {@link NearCacheRecordStore}.
  *
- * @param <V> the type of the value stored by this {@link NearCacheRecord}
+ * @param <V> the type of the value
+ *            stored by this {@link NearCacheRecord}
  * @see com.hazelcast.internal.eviction.Expirable
  * @see com.hazelcast.internal.eviction.Evictable
  */
 public interface NearCacheRecord<V> extends Expirable, Evictable<V> {
 
+    /**
+     * Base time to be used for storing time values as diffs
+     * (int) rather than full blown epoch based vals (long)
+     * This allows for a space in seconds, of roughly 68 years.
+     *
+     * Reference value (1514764800000) -
+     * Monday, January 1, 2018 12:00:00 AM
+     *
+     * The fixed time in the past (instead of {@link
+     * System#currentTimeMillis()} prevents any time
+     * discrepancies among nodes, mis-translated as diffs
+     * of -1 ie. {@link Record#UNSET} values. (see.
+     * https://github.com/hazelcast/hazelcast-enterprise/issues/2527)
+     */
+    long EPOCH_TIME = zeroOutMs(1514764800000L);
+
     int TIME_NOT_SET = -1;
 
     long NOT_RESERVED = -1;
-    long RESERVED = -2;
```

Comment:
Can you give some background as to how the updated coordination happens now? I assume now concurrent access is guarded by the reservation ID. Can you elaborate the lifecycle and how the reservation ID prevents concurrent access?

Reply:
1. Before filling near cache with actual value, first we put a record for reservation.
2. then we do remote call
3. when remote call returns, we check if it is the same reservation record before we do remote call
4. then we put the returned value to near cache or discard

We need this because during these steps, an invalidation may come in parallel from server. If invalidation comes in parallel, it deletes reserved record. Hence at step 3, we can decide not to put remote value into near cache.

Reply:
Thanks!